### PR TITLE
Add Ibus telemetry

### PR DESCRIPTION
**Pull request by CraigJPerry**
Hey Guys,

I've added ibus telemetry support and i've tested it on an STMF3DISCOVERY and a NAZE32 clone, both speaking to an FS-IA6B rx with FS-I6 tx.

Can i get some others to also test please. Here's a naze image (cleanflight master / 1.12): [cleanflight_NAZE.zip](https://github.com/cleanflight/cleanflight/files/104026/cleanflight_NAZE.zip)

Please test without props, don't fly this one just yet. I don't want any crashed quads on my conscience! :-) Once i've got a few test feedbacks i'll push a flyable version.

Can you please report the temperature reported on your boards please. Ibus has an offset for the temp field which i've slid along to zero, however neither of my boards produce a gyro temperature and i don't have a baro fitted. I'd like to know if there are scaling issues with reported temps.

Cheers,

### CODE REVIEW

**Comment by BadLambdaJamma** [comment]
I am scratch building with a Naze32 Rev 6 B and the Turnigy/FlySky TGY-i10 w/ ia10 rec. and ibus expansion and telemetry modules.  If this would be a meaningful test, let me know.

**Comment by CraigJPerry** [comment]
An i10 yeah that would be good to test with. Could you also post a pic of the i10 telemetry screen please when you get a chance.

Cheers

**Comment by CraigJPerry** [comment]
Pushed an update making ibus sharable with MSP on a serial port.

@cklam2 reported that IBUS wasn't shareable with MSP. I had deliberately left it out of the MSP shareable functions mask because IBUS doesn't shut up. Even if there's no one talking to the receiver it will chat away with the same 4 byte query every 8ms on the serial tx line.

This shouldn't cause a problem for cleanflight on the flight controller, but it may hamper the configurator talking via the msp. It'd be easily worked around i suppose in the basic case where the pilot is connected via usb to their computer - just unplug the battery to drop rx power. Often though you want the rx connected when in the configurator (testing modes, adjustment configs etc.)

@cklam2 offered to test this so i'll ask him to specifically look for problems using the configurator.

**Comment by BadLambdaJamma** [comment]
+1 internets.  We absolutely need IBUS RX shareable with MSP if possible. This will allow people to keep the GPS on the standard pin 3/4 and not resort to softserial for GPS. "4 byte query every 8ms on the serial tx line."  This sounds like a heartbeat to the ground radio?  Should it just be "piped" back "down" via the telemetry pin ? (either external to or by the FC ?).  Can someone link to the IBUS protocol standards or whatever the tribal knowledge source is for these ?

**Comment by BadLambdaJamma** [comment]
I am a few days away from my test as I run out of pin headers to power the 5Volt on the Naze32.  I will be picking them up at the local shop at lunch today.

**Comment by CraigJPerry** [comment]
No worries @BadLambdaJamma just whenever you have time. I'm interested to see the i10 telem screen too.

I reached out to the guys at RCG to see about adding some new telem fields or renaming ext voltage as just "data" but not had anything back yet.

**Comment by BadLambdaJamma** [comment]
What the best place for me to pull all these changes from?  Do you have a naze32 hex ?  I may be able to test tomorrow.

**Comment by CraigJPerry** [comment]
Yeah heres one i made earlier: https://github.com/cleanflight/cleanflight/files/105750/cleanflight_NAZE.zip

I was trying to see if I can publish the URL to the builds that Travis ci produced - then we'd automatically have a hex for every platform but it seems they don't get persisted anywhere.

**Comment by BadLambdaJamma** [comment]
Off Topic a bit - You mentioned the Cleanflight build system.  When looking at ibus or (insert your serial RX Bus here).  It is obvious that only one given protocol will be used for a single flight controller configuration.  With codespace and memory being as tight as it is, this is screaming "MODULAR BUILDS ON DEMAND PLEASE". Basically I want to go to a web page and select my hardware and have a hex built for it.  So my ibus hex does not have any sbus code in it.  The example of serial bus selection is the obvious one for such an on demand build system.   But it could go deeper (turtles all the way down), to include what chipsets are available on the hardware.  By scanning the available hardware, drivers, and selected options, you would get a custom hex.  I know that the unit testing permutations are many but meh, automation rules over that labor.  What do you think?

**Comment by BadLambdaJamma** [comment]
ohhhhh, I forgot, thanks for the hex.  I may be able to test tonight.  still soldering ;-)

**Comment by CraigJPerry** [comment]
Yeah the build already cuts useless parts out based on the target platform already. Doesn't take it as far as your idea but the current situation trims enough that it all fits in the limited memory and changes the end user might make are runtime configurable rather than needing to custom compile a new hex.

It's easier from a project support point of view to have less firmwares - less potential for build bugs. They are super evil.

**Comment by BadLambdaJamma** [comment]
Ok,   I have everything ready to go.  I am using the Naze32 Rev 6B and the turnigy 1a10 receiver.  This one gets wired a little different from the iA6B. Instead of dupont 3-pin connectors for sensors-in and servos-out, there are two mini connectors on the side labelled in and out.  I plugged my telemetry sensors into each other, then into the "input" on the IA10.   The "output" of the ia10 is wired to pin 4 of the Naze32.  I tested this in baseflight and have proper radio operation.  RSSI, TX batt, RX batt, External temp, and external battery probe telemetry is all displayed nicely on the i10 radio.  I am not sure how to feed telemetry into the Naze32 at this point.  How should telemetry be wired?  (I have soldered pin headers on uart #1 so it available to swap with MSP)

**Comment by CraigJPerry** [comment]
Telemetry comes from the naze (tx pin of your chosen UART) directly to the IBUS in port. It doesn't make sense to add the weight of those sensors just to dupliucate the data the naze already has anyway (temperature reading from the gyro, battery voltage from vbat). When you connect the naze direct to the ia10 you should see 3 virtual sensors - battery voltage, temperature and throttle (motor rpm).

**Comment by BadLambdaJamma** [comment]
The derp is hard.  Sorry for being daft.  Assuming that the IBUS telemetry output from the Naze32 can be chained  through the turnigy supplied telemetry "sensors"???  Specifically may want to use the turnigy temp probe with the Naze32 built in values.

**Comment by hydra** [comment]
any update on testing this?

**Comment by hopoffbaby** [comment]
Hi,

Is there a build of this for SPRacingF3 that I can test?

Thanks

**Comment by inode-** [comment]
Can you give us more information about the Flysky i6 setup to receive the telemetry data?

**Comment by CraigJPerry** [comment]
Hello, See the commentary in ibus.c for an in-depth explanation of the
protocol.

**Comment by unitware** [comment]
Hi, I'd like to test this with Flysky FS-i6 transmitter, FS-iA6B receiver and a Naze32 rev6A
I have flashed the naze but I do not see the iBus option in the ports tab, for the telemetry function when trying to configure with cleanflight configurator 1.2.2.

Is there any CLI commands or similar that I should know of (I'm a rookie on QUADS, not so rookie in software development)

**Comment by CraigJPerry** [comment]
Hey @unitware many thanks for posting the screenshot, I've been hoping for a few more captures like this.

Seems the gyro temperature isn't available on any naze or naze derivatives. I remember looking up the spec sheet for my gyro and it said temperature was available. Ho hum. At least im not alone.

The motor reading is showing throttle position, 1150 is just what you've set your min throttle to (seems a tiny bit high? My quad would be skyward bound at 1150 lol)

Thanks for this!

**Comment by unitware** [comment]
Hi, i created a fork of cleanflight, pulled your branch and rebased it to v1.13.0-RC1

Somehow I could not create a pull request to your branch but you can see the result here:
https://github.com/unitware/cleanflight/tree/ibus-telemetry

Minor changes really, use it but nice to stay tuned to the main development.

**Comment by sinokop-maksim** [comment]
HI.
on the new FlySky works fine.

what about virtual sensors? GPS Sats,Speed,ALT... ?
img_20160529_143740
img_20160529_143747

**Comment by DzikuVx** [comment]
@CraigJPerry I just ported it to iNavFlight#259 
Thank you very much for this code!

**Comment by sarpba** [comment]
I tryed @CraigJPerry Cleanflight Naze32 and @DzikuVx iNav spraceingF3 firmwares to. I experienced the same error. The RX and TX randomly disconected, when the FC inject the telemetry data into the RX. It works befor disconect I can se the telemetry dada from my FC on my TX.

Config:
SP Raceing F3 or Naze32 FC
Turnigy TGY-I10 (1.0.7.0 or 1.0.8.0 firmware) transmitter
FS-IA6B receiver

Someone else? Comes out of this problem?

**Comment by mojocorp** [comment]
Where are the instructions and the connection diagram ?
I'm already using ibus (RX_SERIAL + IBUS) by connecting the UART2 RX Pin to the FlySky iBus 'servo' connector. Do I just need to activate iBus telemetry ? Do I need to connect UART2 TX to iBus 'sensor' ?
Is it possible to use iBus as rx and at the same time the ibus telemetry on a single UART ?

**Comment by unitware** [comment]
@mojocorp I think you need two uarts in that case, I'm using ppm from the tx and then uart 2 TX to the ibus sensor pin. It's configured as bidirectional so both tx and rx is used. Please correct me if I'm wrong because I'd love to use serial_rx as well but didn't think it was worth another uart.

**Comment by mojocorp** [comment]
@unitware thanks for the info. I've reconnected my receiver as PPM and connected UART2 TX to the receiver 'sensor' and now it's working. yuhu. I had to patch the flight configurator to add the ibus telemetry option. Now I'm playing with the source code and will try to integrate the altitude.

**Comment by unitware** [comment]
Yes the option was missing in the flight configurator, I used CLI to enable the ibus function. 

I have a branch which I have rebased on v1.13 and currently I'm working on some unit tests. Also I did divide the vbat with the number of cells since my TX does not support setting alarm levels above 10 Volt. 

(https://github.com/unitware/cleanflight/tree/ibus-telemetry)

**Comment by mojocorp** [comment]
@CraigJPerry The temperature reported on my Naze32 is 0, but it's due to telemTemperature1, not from the telemetry code.
The temperature seems to be encoded as 10th of degrees with the origin at -40. If I use the barometer, I get 38.6C on my display.

#ifdef BARO
        float temperature = (baroTemperature + 50) / 100.f; // this comes from frsky telemetry
#else
        float temperature = telemTemperature1 / 10;
#endif
        //temperature should be float, expressed in 1/10 degree.
        sendIbusMeasurement(address, (uint16_t) ((temperature + 40)*10));

**Comment by BadLambdaJamma** [comment]
unitware if you need an ibus telemetry sensor to continue developing this code, I can send you one.  (voltage, temp or RPM).

**Comment by unitware** [comment]
@BadLambdaJamma Thanks, I can see in your post above that you want to daisy chain the sensors and that should be possible if the naze is in the end. At the moment the adresses are hardcoded but that should not be the case. If I find the time than I will implement that functionality in a few weeks from now but not sooner. (other stuff has priority at the moment) 

Thanks for the offer of sending a sensor. I should be able to do that implementation without hardware if you are willing to test it later on. (or someone else, @mojocorp seems active at the moment).

**Comment by caveymon** [comment]
Hey there! I've recently tried out your hex file from January. I succesfully got VBat as Ext2. I had 0 reading on Tem3. Not much feedback this way, but I thought I'd chime in. I would really like to see this make it's way into the main branch.

My setup: FS-I6 with FS-IA6B, stock firmware running through Naze32 rev5.

**Comment by unitware** [comment]
@BadLambdaJamma @CraigJPerry 

I have a question about the daisy chaining:
My assumption is that a sensor will only relay commands to addresses that it's not using. Thus the lowest address that a sensor sees on the bus will be the one to use, and forward the rest down the chain. Can any of you confirm this? (there are multiple other ways to do it, using dip switches etc.)

```
*  /---------\
 *  | IBUS RX | > Hello sensor at address 1, are you there?
 *  \---------/     [ 0x04, 0x81, 0x7A, 0xFF ]
 *
 * Due to the daisy-chaining, this hello also serves to inform the sensor
 * of its address (position in the chain). There are 16 possible addresses
 * in iBus, however address 0 is reserved for the rx's internally measured
 * voltage leaving 0x1 to 0xF remaining.
 *
 * Having learned it's address, the sensor simply echos the message back:
```

**Comment by unitware** [comment]
@caveymon If you want you may try this out, it's based on the v1.13 release and includes the baro temperature that @mojocorp suggested above.

Dropbox link: https://www.dropbox.com/s/a49qv02ty0y5fgj/cleanflight_NAZE.hex?dl=0

**Comment by caveymon** [comment]
@unitware @mojocorp I'll give it a go this evening!

**Comment by caveymon** [comment]
Alright, gave it a go, and stumbled upon some stuff.

The BARO temperature seems to be 10 degrees above ambient temperature. I got measurements of 35 degrees centigrade, when ambient temperature was about 25 (is this expected behaviour?)

The VBat information seemed to work at first. I got 7.7V and 11.2V from a 2S resp. 3S battery, but after setting up the Naze to react properly to my switches the VBAT reported by CleanFlight and through ibus telemetry mismatched. Cleanflight (and the status command) show a proper 7.7V, whereas the telemetry reports 1S voltages of 3.85V. This triggers the battery alarm on the transmitter.

The Throttle control works as expected. It showed my initial throttle position (1050) and the top end reported as 1996. Which corresponds to what Cleanflight reports in the receiver and motor tabs.

**Comment by unitware** [comment]
ok, thanks for the test.

The voltage is (almost) as expected, In this version I have set it up to report calculated cell voltage rather than the actual Vbatt measurement. I prefer it this way mostly because that I cannot set battery warning levels higher than 10 volt in my tx (FlySky FS-i6). And the tx does not need to be altered to handle different batteries. It's a little odd that you got "correct" readings first, but it will occur if the battery sensor code does not detect the cells properly at startup - maybe because you were powering via USB before connecting the battery? 

Temperature: The ambient temperature should be lower than the board/chip temperature that is actually measured, making it a bit hard to verify properly. That said there can absolutely be an error in the conversion code. We need to investigate the datasheets to verify that it's correct, the code is originating from another telemetry module so it should be ok.

**Comment by caveymon** [comment]
Thanks for the explanation. 

I figured something like that was going on with VBat. I'll give the TX menu another rundown and set the battery warning accordingly. I was powering through USB before connecting the battery, so that might explain the deviation. The consistency right now is 100%, so your code for showing VBat/cell-count as voltage works as expected.

Temperature: ok, the deviation from ambient is to be expected. Thanks for the clarification.

**Comment by mattw1234** [comment]
I have the iBus telemetry output on UART2 from the Tx pin, which pin should I connect to on the FS-iA6B? Is it possible to use the iBus for telemetry and also for the receiver connection (instead of PWM or PPM)? Thanks

**Comment by unitware** [comment]
@mattw1234 what I know you need two different uarts. I'm running ppm for that reason. 

In theory there is a slight chance that it could work to wire smartly but that would rely on the receiver not issuing ibus and sensor commands att the same time. possibly some more firmware work is needed to. I have not dug into that - but I'd really like to have both on the same port.

**Comment by unitware** [comment]
@BadLambdaJamma I have made a hexfile that might work with daisychained sensors (and naze32 in the end) could you try it for me? I have made the assumption that every unit in the chain only resends the adresses not used further down the line, which might be valid.

If it does not work then it would be very helpful to get a serial dump from the signal line at various points in the daisy chain.

(Do not fly this one, I have not. And make a backup as always):
https://www.dropbox.com/s/d8r04bb36318hi2/cleanflight_NAZE_daisy_chainable.hex?dl=0

**Comment by abramovtv** [comment]
Works fine for me with "Sens" connected to pin 3 of UART2 and firmware from zip file.
Are you @CraigJPerry planning merge this code to master?

**Comment by CraigJPerry** [comment]
Not me @Timur (I don't fly CF anymore) but @Magnus has been doing great
work here recently

**Comment by Argosse** [comment]
Does the existing code for Ibus hear the receiver 'e.g. ia6b' ask for the other IBus address' we don't know?

If it does; are we able to spoof as the rest of the 5 or 6 addresses? I would really like to have latitude/longitude displayed on my Flysky i6s from Flip32+ with Neo M8n thru i2c nav board.


 I'm looking at moving from MW on an arduino mega, to a Flip32+.  I have a FlySky i6s and an ia6b receiver [unfortunately stuck in different state for another month or so] and Ibus seems exactly what I'm looking for flying. I plan to add a Neo M8n unit thru an i2c nav board or softserial; whicher ends up easier.

**Comment by cj1324** [comment]
👍 

board:  Seriously Pro Racing F3 EVO
Radio: FS-IA10B + Flysky i6s

img_20160918_025723 1

**Comment by unitware** [comment]
Would you like to test my branch as well? It's merged to current master, unit tested and with a best guess support for daisy chained sensors. 

I will create a pull request as soon as someone has been flying it. 



Skickat från min iPhone

**Comment by cj1324** [comment]
@unitware Failed.
process：
+ checkout  unitware/cleanflight/tree/ibus-telemetry
+ make TARGET=SPRACINGF3EVO hex
+ use cleanflight-configurator
+ DFU flash  local firmware 
+ try connect serial (10 sec timeout)
+ red light (Red light flashes four times per round)
+ replace my branch firmware ok.
connect log:
```
2016-09-18 @ 上午9:00:48 -- Running - OS: Windows, Chrome: 54.0.2840.27, Configurator: 1.2.3
2016-09-18 @ 上午9:02:40 -- Serial port successfully opened with ID: 14
2016-09-18 @ 上午9:02:40 -- MultiWii API version received - 1.22.0
2016-09-18 @ 上午9:02:41 -- Flight controller info, identifier: CLFL, version: 1.13.0
2016-09-18 @ 上午9:02:41 -- Running firmware released on: Sep 18 2016 08:47:27
2016-09-18 @ 上午9:02:50 -- No configuration received within 10 seconds, communication failed
```
China CST timezone

How to debug ?

**Comment by unitware** [comment]
Sorry that it failed, did you clear configuration prior to flashing? 

Skickat från min iPhone

**Comment by Argosse** [comment]
@cj1324 Would you mind testing my newest built hex for the evo? I've tried spoofing address 4 in a new way. Thanks for testing!

https://www.dropbox.com/sh/9zcqmjiq5fb8u4f/AAApbraSayhax4azlh8TV9Hha?dl=0

**Comment by cj1324** [comment]
@Argosse Still fails.
```
2016-09-18 @ 下午9:50:04 -- Serial port successfully opened with ID: 3
2016-09-18 @ 下午9:50:05 -- MultiWii API version received - 1.19.0
2016-09-18 @ 下午9:50:05 -- Flight controller info, identifier: CLFL, version: 1.13.0
2016-09-18 @ 下午9:50:05 -- Running firmware released on: Sep 17 2016 08:11:52
2016-09-18 @ 下午9:50:05 -- Board: SPEV, version: 0
2016-09-18 @ 下午9:50:05 -- Unique device ID received - 0x2f00265533571220303834
2016-09-18 @ 下午9:50:16 -- CLI mode detected
```
result:
img_20160918_215440 1


@unitware  use this option?  Still fails.
flash_unit

**Comment by unitware** [comment]
Thanks, I'll have to read up on the changes in master so far and see whats up. 

Skickat från min iPhone

**Comment by cj1324** [comment]
@Argosse
i use  saleae login Analyzer  for you firmware .
img_20160918_224536 1

seleaelogic

found 0x04 0x84 ( power-up 0.8ms)
found

csv data:
[saleae_login_export_csv_data.zip](https://github.com/cleanflight/cleanflight/files/478780/saleae_login_export_csv_data.zip)

**Comment by cj1324** [comment]
@Argosse The old version of the firmware, Protocol data successfully. But not visible
raw

**Comment by unitware** [comment]
@cj1324 @Argosse 
It seems that my rebase previously was not made corectly. I have tried to fix this and the current state of my branch seems to start up at least. (i.e the configurator can connect to the hardware. I do not have access to the radio at this time)
@cj1324: If you have the equipment set up and some time to spare, please try again.
https://github.com/unitware/cleanflight/tree/ibus-telemetry

**Comment by unitware** [comment]
Awesome, thanks. 
I'll create a pull request for this shortly. 

Skickat från min iPhone

**Comment by Argosse** [comment]
Updated hexes, changed address to use builtin address type and the address asked for by ibus instead of 0x04. Thanks for testing !

File: src/main/telemetry/ibus.c [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    return checksum;
+}
+
+static bool isChecksumOk(uint8_t ibusPacket[static IBUS_CHECKSUM_SIZE], size_t packetLength, uint16_t calcuatedChecksum) {
+    // Note that there's a byte order swap to little endian here
+    return (calcuatedChecksum >> 8) == ibusPacket[packetLength - 1]
+            && (calcuatedChecksum & 0xFF) == ibusPacket[packetLength - 2];
+}
+
+static void transmitIbusPacket(uint8_t ibusPacket[static IBUS_MIN_LEN], size_t packetLength) {
```
Comment (ledvinap) [inline src/main/telemetry/ibus.c]: Another possibility is to send checksum without storing it in buffer. Then it won't be necessary to allocate checksum bytes in packet .. but `packetLength` gets a bit confusing

static void transmitIbusPacket(uint8_t ibusPacket[static IBUS_MIN_LEN], size_t packetLength) {
    uint16_t checksum = calculateChecksum(ibusPacket, packetLength);
    for (size_t i = 0; i < packetLength - IBUS_CHECKSUM_SIZE; i++) {
        serialWrite(ibusSerialPort, ibusPacket[i]);
    }
    serialWrite(ibusSerialPort, (checksum & 0xFF));
   serialWrite(ibusSerialPort, (checksum >> 8));
}

**Comment by cj1324** [comment]
@Argosse Sorry, but the receiver burned, buying a new transport.

**Comment by Argosse** [comment]
Ouch. What did I do! Well, we shall try another way. Thanks again for testing.

**Comment by hydra** [comment]
replaced by #2415 - closing.

**Comment by schwabe** [comment]
Just for completeness sake, another bench report for the IBUS telemetry. But with unusual FlySky TX (DIY Multi module + OpenTX)
dsc_1650

**Pull request closed without merging**